### PR TITLE
[docs] Added auth documentation to connect to Amazon Cognito

### PR DESCRIPTION
**Pull request by tbehunin**
# Why

The Expo authentication documentation (https://docs.expo.dev/guides/authentication/) encouraged contributions to the most requested services/providers. Amazon Cognito was a glaring omission.

# How

I leveraged similar examples from the docs, from blogs (https://advancedweb.hu/how-to-secure-the-cognito-login-flow-with-a-state-nonce-and-pkce/), links to sample code in comments on canny (https://expo.canny.io/feature-requests/p/authsession-aws-cognito), and workaround fixes to issues raised (https://github.com/expo/expo/issues/8300).

# Test Plan

- Reviewed and followed steps on how to contribute to documentation
- Updated markdown accordingly
- Save, reload and verify
- Created Cognito icon, matched file specifications of other auth provider icons (256x256 png with transparency)

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by tbehunin** [comment]
@Simek @amandeepmittal @kbrandwijk I wasn't sure where to put a changelog update for a PR with just documentation updates, but apparently the expo-bot doesn't like that to be the root changelog. Can you let me know where I should add this - if at all? Thanks!

File: docs/pages/guides/authentication.mdx [hunk 1]
```diff
@@ -258,6 +264,128 @@ export default function App() {
...
+        <Button
+        title="Logout"
+        onPress={() => {
+            logout();
+            }}
+        />
+      }
+    </>
+  );
+}
```
Comment (Simek) [inline docs/pages/guides/authentication.mdx:380]:
```suggestion
import * as React from 'react';
import * as WebBrowser from 'expo-web-browser';
import { useAuthRequest, exchangeCodeAsync, revokeAsync, ResponseType } from 'expo-auth-session';
import { Button, Alert } from 'react-native';

WebBrowser.maybeCompleteAuthSession();

const clientId = '<your-client-id-here>';
const userPoolUrl =
  'https://<your-user-pool-domain>.auth.<your-region>.amazoncognito.com';
const redirectUri = 'your-redirect-uri';

export default function App() {
  const [authTokens, setAuthTokens] = React.useState(null);
  const discoveryDocument = React.useMemo(() => ({
    authorizationEndpoint: userPoolUrl + '/oauth2/authorize',
    tokenEndpoint: userPoolUrl + '/oauth2/token',
    revocationEndpoint: userPoolUrl + '/oauth2/revoke',
  }), []);

  const [request, response, promptAsync] = useAuthRequest(
    {
      clientId,
      responseType: ResponseType.Code,
      redirectUri,
      usePKCE: true,
    },
    discoveryDocument
  );

  React.useEffect(() => {
    const exchangeFn = async (exchangeTokenReq) => {
      try {
        const exchangeTokenResponse = await exchangeCodeAsync(
          exchangeTokenReq,
          discoveryDocument
        );
        setAuthTokens(exchangeTokenResponse);
      } catch (error) {
        console.error(error);
      }
    };

    if (response) {
      if (response.error) {
        Alert.alert(
          'Authentication error',
          response.params.error_description || 'something went wrong'
        );
        return;
      }
      if (response.type === 'success') {
        exchangeFn({
          clientId,
          code: response.params.code,
          redirectUri,
          extraParams: {
            code_verifier: request.codeVerifier,
          },
        });
      }
    }
  }, [discoveryDocument, request, response]);

  const logout = async () => {
    const revokeResponse = await revokeAsync(
      {
        clientId: clientId,
        token: authTokens.refreshToken,
      },
      discoveryDocument
    );
    if (revokeResponse) {
      setAuthTokens(null);
    }
  };
  console.log('authTokens: ' + JSON.stringify(authTokens));
  return authTokens ? (
    <Button title="Logout" onPress={() => logout()} />
  ) : (
    <Button disabled={!request} title="Login" onPress={() => promptAsync()} />
  );
}
```

Code reformat and few tweaks.

File: docs/pages/guides/authentication.mdx [hunk 2]
```diff
@@ -258,6 +264,128 @@ export default function App() {
...
 
+### Cognito
+
+<CreateAppButton name="Cognito" href="https://us-east-1.console.aws.amazon.com/cognito/v2/idp/user-pools?region=us-east-1" />
+
+| Website                      | Provider  | PKCE          | Auto Discovery |
+| ---------------------------- | --------- | ------------- | -------------- |
+| [Get Your Config][c-cognito] | OpenID    | Supported     | Not Available  |
+
+[c-cognito]: https://us-east-1.console.aws.amazon.com/cognito/v2/idp/user-pools?region=us-east-1
```
Comment (Simek) [inline docs/pages/guides/authentication.mdx:275]:
```suggestion
[c-cognito]: https://console.aws.amazon.com/cognito/v2/idp/user-pools
```

File: docs/pages/guides/authentication.mdx [hunk 3]
```diff
@@ -258,6 +264,128 @@ export default function App() {
...
+[c-cognito-api-docs]: https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-userpools-server-contract-reference.html
+
+- Leverages the Hosted UI in Cognito ([api documentation][c-cognito-api-docs])
```
Comment (Simek) [inline docs/pages/guides/authentication.mdx:278]:
```suggestion
- Leverages the Hosted UI in Cognito ([API documentation][c-cognito-api-docs])
```

Case fix.

**Review by Simek — COMMENTED** [review]
Thank you for adding this guide, this looks nice! 👍 

I have left few small comments, plus updated and reformatted example code, but with those changes in we probably can ship this.

**Comment by tbehunin** [comment]
@Simek Thanks for the suggestions! I applied them all. Let me know if there's anything else I should do here.

**Review by Simek — APPROVED** [review]
Thank you for the quick review updates! 🙏 LGTM, let's ship this!

**Pull request merged**